spec/data-model.md

```diff
@@ -0,0 +1,177 @@
+# DRAFT MessageFormat 2.0 Data Model
+
+To work with messages defined in other syntaxes than that of MessageFormat 2,
+an equivalent data model representation is also defined.
+Implementations MAY provide interfaces which allow
+for MessageFormat 2 syntax to be parsed into this representation,
+for messages presented in this representation to be formatted,
+or for other operations to be performed on or with messages in this representation.
+
+Implementations are not required to use this data model for their internal representation of messages.
+
+To ensure compatibility across all platforms,
+this interchange data model is defined in terms of JSON-compatible values.
+While this document uses TypeScript syntax for their definition,
+the canonical and authoritative source is the `message.json`
+[JSON Schema](https://json-schema.org/) definition.
+
+## Messages
+
+A `SelectMessage` corresponds to a syntax message that includes _selectors_.
+A message without _selectors_ and with a single _pattern_ is represented by a `PatternMessage`.
+
+```ts
+type Message = PatternMessage | SelectMessage
+
+interface PatternMessage {
+  type: 'message'
+  declarations: Declaration[]
+  pattern: Pattern
+}
+
+interface SelectMessage {
+  type: 'select'
+  declarations: Declaration[]
+  selectors: Expression[]
+  variants: Variant[]
+}
+```
+
+Each message _declaration_ is represented by a `Declaration`,
+which connects the left-hand side `target` _variable_
+with its right-hand side `value`.
+
+```ts
+interface Declaration {
+  target: VariableRef
+  value: Expression
+}
+```
+
+In a `SelectMessage`, the `keys` and `value` of each _variant_ are kept as an ordered list.
+For the `CatchallKey`, a string `value` may be provided to retain an identifier.
+This is always `'*'` in MessageFormat 2 syntax, but may vary in other formats.
+
+```ts
+interface Variant {
+  keys: Array<Literal | CatchallKey>
+  value: Pattern
+}
+
+interface CatchallKey {
+  type: '*'
+  value?: string
+}
+```
+
+## Patterns
+
+Each `Pattern` represents a single linear _pattern_ without selectors,
+with a `body` made up of `Text` and `Expression` shapes.
+`Text` represents literal _text_,
+while `Expression` wraps each of the potential _expression_ shapes.
+The `value` of `Text` is the "cooked" value (i.e. escape sequences are processed).
+
+Implementations MUST NOT rely on the set of `Expression` `body` values being exhaustive,
+as future versions of this specification MAY define additional expressions.
+If encountering a `body` with an unrecognised value,
+an implementation SHOULD treat it as it would a `Reserved` value.
```

Comment:
This is required to deal with changes that may arise in MessageFormat 2.1 or later, if we do choose to make some use of the reserved characters.

Let's say we add inline comments with a syntax like `{# ... #}`. If this happens, we'll want to also be able to represent these in the data model, where they'll show up in the `Expression` body. This sentence lets us do that so that a MessageFormat 2.0 processor won't change its behaviour if it's fed a MessageFormat 2.1 data model.